File: Example/BLECentralManager_iOS/ControlCharacteristic.m
//
//  ControlCharacteristic.m
//  BLECentralManager_iOS
//
//  Created by Balázs Kilvády on 8/1/16.
//  Copyright © 2016 Balázs Kilvády. All rights reserved.
//

#import "ControlCharacteristic.h"

@implementation ControlCharacteristic

- (void)device:(BLECDevice *)device didFindCharacteristic:(CBCharacteristic *)characteristic
{
    DLog(@"device control characteristic %@ found!", characteristic.UUID);
    if (device.peripheral) {
        [device.peripheral readValueForCharacteristic:characteristic];
    }
}

- (void)device:(BLECDevice *)device
didUpdateValueForCharacteristic:(CBCharacteristic *)characteristic
         error:(NSError *)error
{
    if (characteristic.value == nil) {
        DLog(@"No value!?");
        return;
    }

    const uint8_t *bytes = (const uint8_t *)[characteristic.value bytes];
    [_delegate controlDidUpdate:bytes[0] == 0 ? ButtonActionStart : ButtonActionStop];
}

@end

File: Example/BLECentralManager_iOS_Swift/ControlCharacteristic.swift
//
//  ControlCharacteristic.swift
//  BLECentralManager_iOS_Swift
//
//  Created by Balázs Kilvády on 7/26/16.
//  Copyright © 2016 kil-dev. All rights reserved.
//

import BLECentralManager
import CoreBluetooth


class ControlCharacteristic: BLECCharacteristicDelegate {

    weak var delegate: ControlCharacteristicDelegate?

    func device(_ device: BLECDevice,
                didFindCharacteristic characteristic: CBCharacteristic) {
        DLog("device control characteristic <\(characteristic.uuid)> found!")
        device.peripheral?.readValue(for: characteristic)
    }

    func device(_ device: BLECDevice,
                didUpdateValueForCharacteristic characteristic: CBCharacteristic,
                error: Error?) {
        guard let data = characteristic.value else {
            DLog("No value!?")
            return
        }

        let array = data.withUnsafeBytes {
            [UInt8](UnsafeBufferPointer(start: $0, count: data.count))
        }
        let byte: UInt8 = array[0]
        delegate?.controlDidUpdate(byte == 0 ? .start : .stop)
    }
}
